content/learn/intermediate/control-flow-2.md

```diff
@@ -0,0 +1,268 @@
+
+  ```go
+  {{ $empty := cslice }}
+  {{ range $empty }}
+      {{/* ... */}}
+  {{ else }}
+      slice was empty
+  {{ end }}
+  ```
+
+#### Accessing global context data in range
+
+The following program illustrates a common error for first-time users of `range`.
+
+```go
+{{ $nums := cslice 1 2 3 }}
+{{ range $nums }}
+    {{/* ... */}}
+    {{ .User.Username }} {{/* ERROR: can't evaluate field User in type inteface {} */}}
+{{ end }}
+```
+
+The problem is that, inside the range block, the dot `.` is overwritten by successive elements of the slice `1`, `2`,
+`3`. While this behavior is generally useful—we often _want_ to refer to the current element in a range action—it is
+counterproductive here, as `.User.Username` tries to look up the field `User` on an integer (and fails to do so.) What
+we really want is to access the global context data as it was before the range loop. One solution is to save the
+original context data in a variable prior to the loop:
+
+```go
+{{ $dot := . }}
+{{ range ... }}
+    {{ $dot.User.Username }}
+{{ end }}
+```
+
+To make this pattern easier, before each custom command execution, YAGPDB predefines the variable `$` as the initial
+context data for you.
+
+{{< callout context="caution" title="Warning: Common Error" icon="outline/alert-triangle" >}}
+
+In a range block, the dot is overwritten by elements of the slice or map, so code such as `.User.Username` is likely to
+error. If you need to access the global context data, do so through the predefined `$` variable instead.
+
+```go
+{{ range ... }}
+    {{ $.User.Username }} {{/* instead of .User.Username */}}
+{{ end }}
+```
+
+{{< /callout >}}
+
+### While
+
+`while` loops as long as the specified condition is truthy. Unlike the `range` action, the dot `.` is not affected.
+
+For instance, the following code loops as long as `$n` is not 1. In each iteration, `$n` is updated to either `n/2` or
+`3n+1`.
+
+```go
+{{ $n := 19 }}
+
+{{ print $n " " -}}
+{{ while ne $n 1 }}
+    {{- if eq (mod $n 2) 0. }}
+        {{- $n = div $n 2 }}
+    {{- else }}
+        {{- $n = mult $n 3 | add 1 }}
+    {{- end -}}
+    -> {{ print $n " " -}}
+{{ end }}
+```
+
+Many `while` loops can be written as a more idiomatic range loop instead. In particular, to iterate a fixed number of
+times, use `{{ range n }}` as in `{{ range 5 }}` instead of maintaining your own counter variable with `while`.
+
+### Break and Continue
+
+In custom commands, we provide two actions to control the flow of loops: `{{ break }}` and `{{ continue }}`. `break`
+exits the loop prematurely, whereas `continue` skips the remainder of the current iteration and jumps to the next one.
+These can prove very useful to optimize your code for size and readability, with similar benefits to guard clauses with
+`{{ return }}` introduced in earlier chapters.
+
+## With Blocks
+
+Just like the `if` action, `with` runs a block of code if the given expression is truthy. The only difference is that
```

Comment:
I don't think truthiness is explained anywhere, this may not be the best place but it should be clarified somewhere.

Reply:
Agreed, but finding a home for this information is surprisingly difficult. The most natural location—Control Flow 1 introducing if conditionals—is not totally appropriate, unfortunately, since the complex data types have not been introduced at that stage. I suppose we could only show the boolean case in Control Flow 1 and allude to if conditionals operating on a more general definition of empty values. The empty values for each data type could then be clarified as they are introduced in Data Types 1 and Data Types 2 (so, for instance, after presenting the slice type we mention that an empty slice is falsy.)

A true conundrum indeed.

Reply:
I think interspersing it with `if`-actions and the respective data-type is a sensible approach, because there really isn't any good place to introduce this in a centralised paragraph.
~~I also have the hope that future readers will actually read rather than skim but that's a pipe dream of mine~~

Reply:
That sounds like a good approach to me too, the initial explanation could also be expanded upon to include numbers and strings along with booleans.

Reply:
Done in https://github.com/botlabs-gg/yagpdb-docs-v2/pull/41/commits/916648171a27b96e830eddf631ef4fff5c9a3825.